File: js/src/engine/tile-map.js
'use strict';

import { Dimension, Rectangle } from './struct';
const [COLLISIONS, IMAGES, LOADED, PAGES] = [Symbol(), Symbol(), Symbol(), Symbol()];

// a TileMap provides an efficient way to store and render a static tile map
// and also calculate wall collisions quickly
class TileMap {
  [IMAGES] = [];
  [COLLISIONS] = [];
  [PAGES] = [];

  constructor(tm, {meta, images, collisions}) {
    this[PAGES] = meta.pages;
    this.tw = meta.tw;
    this.th = meta.th;
    const tile = (n) => {
      for(let page of meta.pages) {
        if(page.min <= n && n < page.max) {
          n -= page.min;
          const tp = tm.pages[page.name];
          if(!tp) { return null; }
          const src = new Rectangle(this.tw * (n % (tp.width / this.tw)), this.th * Math.floor(n / (tp.width / this.tw)), this.tw, this.th);
          return [tp, src];
        }
      }
    }
    for(let depth of Object.keys(images)) {
      const dest = new Rectangle(0, 0, this.tw, this.th);
      const layer = document.createElement('CANVAS');
      layer.width = images[depth][0].length * this.tw;
      layer.height = images[depth].length * this.th;
      const ctx = layer.getContext('2d');
      this[IMAGES].push([+depth, layer]);
      for(let row of images[depth]) {
        for(let n of row) {
          const t = tile(+n);
          if(t) {
            const [tp, src] = t;
            ctx.drawImage(tp, ...src, ...dest);
          }
          dest.x += this.tw;
        }
        dest.x = 0;
        dest.y += this.th;
      }
    }
    this[COLLISIONS] = collisions.map(l => l.split('').map(i => !!+i));
  }

  get pages() {
    return this[PAGES].map(({name}) => name);
  }

  get size() {
    return new Dimension(this.tw * this[COLLISIONS][0].length, this.th * this[COLLISIONS].length);
  }

  // copy the tile map to the main canvas
  draw(draw) {
    for(let [depth, image] of this[IMAGES]) {
      draw.image(image, [0, 0, image.width, image.height], [0, 0, image.width, image.height], depth);
    }
  }

  // check if a given Rectangle collides with the collision map
  collides(box) {
    box = [...box];
    box[2] = Math.ceil((box[0] + box[2]) / this.tw);
    box[3] = Math.ceil((box[1] + box[3]) / this.th);
    box[0] = Math.floor(box[0] / this.tw);
    box[1] = Math.floor(box[1] / this.th);
    for(let i = box[0]; i < box[2]; ++i) {
      for(let j = box[1]; j < box[3]; ++j) {
        if(this[COLLISIONS][j] && this[COLLISIONS][j][i]) { return true; }
      }
    }
    return false;
  }
}

export default TileMap;

File: c++/src/engine/tilemap.cc
#include "tilemap.h"
#include "texture-manager.h"

#include <cmath>
#include <fstream>
#include <sstream>
#include <json.hpp>
using nlohmann::json;

namespace Game {
    TileMap::TileMap(const std::string & file, TextureManager & texture) : _tm{ texture } {
        std::ifstream fin{ file };
        json data;
        fin >> data;
        _tw = data["meta"]["tw"];
        _th = data["meta"]["th"];

        json & pages = data["meta"]["pages"];
        for(unsigned int i = 0; i < pages.size(); ++i) {
            json & obj = pages[i];
            _texture_pages.emplace_back(obj["name"].get<std::string>());
            _indexmap.emplace_back(obj["min"], obj["max"]);
        }

        for(const std::string & row : data["collisions"]) {
            _collisions.emplace_back();
            for(const char & c : row) {
                _collisions.back().emplace_back(c == '1');
            }
        }
        json images = data["images"];
        for(auto layer = images.begin(); layer != images.end(); ++layer) {
            std::istringstream sin{ layer.key() };
            int depth;
            sin >> depth;
            _images[depth] = layer.value().get<std::vector<std::vector<int>>>();
        }
    }

    const std::vector<std::string> & TileMap::textures() const {
        return _texture_pages;
    }

    Dimension TileMap::size() const {
        return { static_cast<int>(_collisions[0].size()) * _tw, static_cast<int>(_collisions.size()) * _th };
    }

    bool TileMap::collides(const Rectangle & where) const {
        auto s = size();
        const int
            w = s.w,
            h = s.h,
            l = std::max(0, where.x / _tw),
            t = std::max(0, where.y / _th),
            r = std::min(static_cast<int>(std::ceil((where.x + where.w) / _tw)), w),
            b = std::min(static_cast<int>(std::ceil((where.y + where.h) / _th)), h);
        for(int i = t; i < b; ++i) {
            for(int j = l; j < r; ++j) {
                if(_collisions[i][j]) {
                    return true;
                }
            }
        }
        return false;
    }

    std::pair<TexturePage *, Rectangle> TileMap::tile_from_index(int n) const  {
        Rectangle where{ 0, 0, _tw, _th };
        int ind = -1;
        for(unsigned int i = 0; i < _indexmap.size(); ++i) {
            auto & range = _indexmap[i];
            if(range.first <= n && range.second > n) {
                n -= range.first;
                ind = i;
                break;
            }
        }
        if(ind == -1) {
            return {nullptr, where};
        }
        auto & texture = _tm.texture(_texture_pages[ind]);
        auto s = texture.size();
        where.x = _tw * (n % (s.w / _tw));
        where.y = _th * (n / (s.w / _tw));
        return {&texture, where};
    }

    void TileMap::render(SDL_Renderer * ren) {
        auto s = size();
        Rectangle where{0, 0, _tw, _th};
        for(auto layer = _images.begin(); layer != _images.end(); ++layer) {
            where.x = where.y = 0;
            std::unique_ptr<SDL_Texture, decltype(& SDL_DestroyTexture)> image {
                SDL_CreateTexture(ren, SDL_PIXELFORMAT_RGBA8888, SDL_TEXTUREACCESS_TARGET, s.w, s.h),
                SDL_DestroyTexture
            };
            SDL_SetRenderTarget(ren, image.get());
            SDL_SetTextureBlendMode(image.get(), SDL_BLENDMODE_BLEND);
            SDL_SetRenderDrawColor(ren, 0, 0, 0, 0);
            SDL_RenderClear(ren);
            for(auto & row : layer->second) {
                for(int col : row) {
                    auto tex = tile_from_index(col);
                    if(tex.first) {
                        SDL_Rect from = tex.second, to = where;
                        SDL_RenderCopy(ren, *tex.first, &from, &to);
                    }
                    where.x += _tw;
                }
                where.x = 0;
                where.y += _th;
            }
            _layers.emplace(layer->first, std::move( image ));
        }

        SDL_SetRenderTarget(ren, nullptr);
    }

    void TileMap::draw(Draw & draw) const {
        auto s = size();
        Rectangle src{0, 0, s.w, s.h};
        for(auto & layer : _layers) {
            draw.image(layer.second.get(), src, src, layer.first);
        }
    }
}
